[libpas] Use madvise(MADV_ZERO) instead of mmap to zero memory

Achierius opened the pull request:
#### 728128cdbd4f5af9f66d7b353bbec180745c706c
<pre>
[libpas] Use madvise(MADV_ZERO) instead of mmap to zero memory
<a href="https://bugs.webkit.org/show_bug.cgi?id=294583">https://bugs.webkit.org/show_bug.cgi?id=294583</a>
<a href="https://rdar.apple.com/153560968">rdar://153560968</a>

Reviewed by Yusuke Suzuki.

When zeroing sufficiently large regions in libpas, we currently use mmap
to make the OS give us new pages (instead of having to memset them
ourselves). On Darwin, however, this approach is pessimizing in many
cases (e.g. the page is still in lazy-zero-fill state, and so notionally
could just be left in-place), and moreover causes issues with
fragmentation inside the kernel that we want to avoid.
The better alternative is to use madvise(p, sz, MADV_ZERO) whenever
possible.

This is not supported on Linux or older Darwin versions, so we will need
to check for the presence of MADV_ZERO. Moreover, madvise can sometimes
fail (e.g. for CoW pages -- though that in particular should not
happen for us), so we need to still support mmap as a fallback.
So in total, there are three places where we may fall back onto mmap:
 1. At compile time, if MADV_ZERO is not defined;
 2. At runtime, if on the first time madvise(MADV_ZERO) was called
    it returned ENOTSUP and thus the latch is set to skip madvise
    thereafter;
 3. At runtime, if for any particular call to madvise(MADV_ZERO) it
    returns an error (e.g. if we use it on a CoW-page).

* Source/bmalloc/libpas/src/libpas/pas_page_malloc.c:
(pas_page_malloc_zero_fill_latch_if_madv_zero_is_supported):
(pas_page_malloc_zero_fill):

Canonical link: <a href="https://commits.webkit.org/296321@main">https://commits.webkit.org/296321@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f42b443bc8821990a4beb4fa9f1d59ac68788321

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108096 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27759 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18179 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113306 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58596 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28453 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36309 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82065 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111044 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22545 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97383 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62497 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21960 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15519 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58051 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/100692 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91903 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15580 "Found 2 new test failures: imported/w3c/web-platform-tests/url/IdnaTestV2.window.html imported/w3c/web-platform-tests/url/toascii.window.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116431 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/106657 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35161 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25887 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91092 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35536 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93660 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90887 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23174 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35783 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13545 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/30985 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35062 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/130946 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34797 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35550 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38155 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36463 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_page_malloc.c:
```diff
@@ -201,6 +216,32 @@ pas_page_malloc_try_allocate_without_deallocating_padding(
+
+    int rc = madvise(base, size, MADV_ZERO);
+    if (rc)
+        madv_zero_supported = (errno != ENOTSUP);
```
were you able to test the case where COW resulted in a different errno? Or is that the usual case?

Achierius replied:
Was able to test (via injection)

dhecht reviewed (commented):
LGTM (non-binding)

Achierius committed:
[libpas] Use madvise(MADV_ZERO) instead of mmap to zero memory
https://bugs.webkit.org/show_bug.cgi?id=294583
rdar://153560968

Reviewed by Yusuke Suzuki.

When zeroing sufficiently large regions in libpas, we currently use mmap
to make the OS give us new pages (instead of having to memset them
ourselves). On Darwin, however, this approach is pessimizing in many
cases (e.g. the page is still in lazy-zero-fill state, and so notionally
could just be left in-place), and moreover causes issues with
fragmentation inside the kernel that we want to avoid.
The better alternative is to use madvise(p, sz, MADV_ZERO) whenever
possible.

This is not supported on Linux or older Darwin versions, so we will need
to check for the presence of MADV_ZERO. Moreover, madvise can sometimes
fail (e.g. for CoW pages -- though that in particular should not
happen for us), so we need to still support mmap as a fallback.
So in total, there are three places where we may fall back onto mmap:
 1. At compile time, if MADV_ZERO is not defined;
 2. At runtime, if on the first time madvise(MADV_ZERO) was called
    it returned ENOTSUP and thus the latch is set to skip madvise
    thereafter;
 3. At runtime, if for any particular call to madvise(MADV_ZERO) it
    returns an error (e.g. if we use it on a CoW-page).

* Source/bmalloc/libpas/src/libpas/pas_page_malloc.c:
(pas_page_malloc_zero_fill_latch_if_madv_zero_is_supported):
(pas_page_malloc_zero_fill):

Canonical link: https://commits.webkit.org/296321@main
